File: languages/js/src/entity/proof.ts
import { AnchorNetwork } from "./anchor";
import * as proto from "../bridge/proto/proof";

export class Proof {
  leaves: string[];
  nodes: string[];
  depth: string;
  bitmap: string;
  anchor: ProofAnchor;

  constructor(
    leaves: string[],
    nodes: string[],
    depth: string,
    bitmap: string,
    anchor: ProofAnchor
  ) {
    this.leaves = leaves;
    this.nodes = nodes;
    this.depth = depth;
    this.bitmap = bitmap;
    this.anchor = anchor;
  }

  static fromProto(p: proto.Proof): Proof {
    return new Proof(
      p.leaves,
      p.nodes,
      p.depth,
      p.bitmap,
      ProofAnchor.fromProto(p.anchor!)
    );
  }

  toProto(): proto.Proof {
    return proto.Proof.fromPartial({
      leaves: this.leaves,
      nodes: this.nodes,
      depth: this.depth,
      bitmap: this.bitmap,
      anchor: this.anchor.toProto()
    });
  }
}

export class ProofAnchor {
  anchorID: number;
  networks: AnchorNetwork[];
  root: string;
  status: string;

  constructor(
    anchorID: number,
    networks: AnchorNetwork[],
    root: string,
    status: string
  ) {
    this.anchorID = anchorID;
    this.networks = networks;
    this.root = root;
    this.status = status;
  }

  static fromProto(p: proto.ProofAnchor): ProofAnchor {
    return new ProofAnchor(
      p.anchorId,
      p.networks.map(x => AnchorNetwork.fromProto(x)),
      p.root,
      p.status
    );
  }

  toProto(): proto.ProofAnchor {
    return proto.ProofAnchor.fromPartial({
      anchorId: this.anchorID,
      networks: this.networks.map(n => n.toProto()),
      root: this.root,
      status: this.status
    });
  }
}

File: languages/go/entity/proof.go
package entity

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

type Proof struct {
	Leaves []string
	Nodes  []string
	Depth  string
	Bitmap string
	Anchor ProofAnchor
}

func NewProofFromProto(p *proto.Proof) Proof {
	if p == nil {
		return Proof{}
	}
	return Proof{
		Leaves: p.Leaves,
		Nodes:  p.Nodes,
		Depth:  p.Depth,
		Bitmap: p.Bitmap,
		Anchor: NewProofAnchorFromProto(p.Anchor),
	}
}

func (p Proof) ToProto() *proto.Proof {
	return &proto.Proof{
		Leaves: p.Leaves,
		Nodes:  p.Nodes,
		Depth:  p.Depth,
		Bitmap: p.Bitmap,
		Anchor: p.Anchor.ToProto(),
	}
}

type ProofAnchor struct {
	AnchorID int64
	Networks []AnchorNetwork
	Root     string
	Status   string
}

func NewProofAnchorFromProto(p *proto.ProofAnchor) ProofAnchor {
	if p == nil {
		return ProofAnchor{}
	}
	return ProofAnchor{
		AnchorID: p.AnchorId,
		Networks: MapAnchorNetworksFromProto(p.Networks),
		Root:     p.Root,
		Status:   p.Status,
	}
}

func (p ProofAnchor) ToProto() *proto.ProofAnchor {
	networks := make([]*proto.AnchorNetwork, len(p.Networks))
	for i, network := range p.Networks {
		networks[i] = network.ToProto()
	}

	return &proto.ProofAnchor{
		AnchorId: p.AnchorID,
		Networks: networks,
		Root:     p.Root,
		Status:   p.Status,
	}
}
